Move GetDataDir to ArgsManager

kiminuo opened the pull request:
This PR attempts to contribute to "Remove gArgs" (#21005).

Main changes:

* `GetDataDir()` function is moved to `ArgsManager.GetDataDirPath()`.
* `GetBlocksDir()` function is moved to `ArgsManager.GetBlocksDirPath()`.

kiminuo commented on src/util/system.cpp:
```diff
@@ -767,34 +768,40 @@ const fs::path &GetBlocksDir()
 }
 
-const fs::path &GetDataDir(bool fNetSpecific)
+void GetDataDir(ArgsManager& argsManager, bool fNetSpecific, fs::path* path)
```
note: `std::string datadir` can be passed instead of `ArgsManager` instance.

kiminuo commented on src/util/system.cpp:
```diff
@@ -767,34 +768,40 @@ const fs::path &GetBlocksDir()
-        if (!fs::is_directory(path)) {
-            path = "";
-            return path;
+        *path = fs::system_complete(datadir);
```
Should we get rid of pointers?

kiminuo commented on src/util/system.h:
```diff
@@ -94,6 +95,15 @@ fs::path GetDefaultDataDir();
 const fs::path &GetDataDir(bool fNetSpecific = true);
+
+/**
+ * Get data directory path.
```
note: Maybe mention that empty datadir leads to default datadir?

kiminuo commented on src/util/system.h:
```diff
@@ -94,6 +95,15 @@ fs::path GetDefaultDataDir();
+ * @param      fnetSpecific Append network identifier to the path.
+ * @param[out] data Directory path.
+ */
+void GetDataDir(ArgsManager& argsManager, bool fNetSpecific, fs::path* path);
```
Note: Is signature in form: `void GetDataDir(fs::path* path, ArgsManager& argsManager, bool fNetSpecific = true);` preferred?

kiminuo commented on src/test/util_tests.cpp:
```diff
@@ -48,24 +48,40 @@ BOOST_FIXTURE_TEST_SUITE(util_tests, BasicTestingSetup)
 {
-    ClearDatadirCache();
-    const fs::path dd_norm = GetDataDir();
+    FastRandomContext insecure_rand_ctx;
+    const fs::path datadir_path = fs::temp_directory_path() / "test_common_" PACKAGE_NAME / insecure_rand_ctx.rand256().ToString();
```
note: Would be nice to wrap this in a helper function.

kiminuo commented on src/test/util_tests.cpp:
```diff
@@ -48,24 +48,40 @@ BOOST_FIXTURE_TEST_SUITE(util_tests, BasicTestingSetup)
 
 BOOST_AUTO_TEST_CASE(util_datadir)
 {
-    ClearDatadirCache();
```
Should the original test be preserved?

laanwj commented:
Concept ACK

hebasto commented:
Concept ACK.

kiminuo commented:
@sipa Yes, you are right. I have modified that.

kiminuo commented on src/util/system.h:
```diff
@@ -94,6 +95,7 @@ fs::path GetDefaultDataDir();
 // The blocks directory is always net specific.
 const fs::path &GetBlocksDir();
 const fs::path &GetDataDir(bool fNetSpecific = true);
+
```
Will remove.

kiminuo commented on src/test/util_tests.cpp:
```diff
@@ -1139,21 +1145,23 @@ BOOST_AUTO_TEST_CASE(util_ReadWriteSettings)
     TestArgsManager args1;
-    args1.LockSettings([&](util::Settings& settings) { settings.rw_settings["name"] = "value"; });
+    args1.ForceSetArg("-datadir", m_path_root.string());
+    args1.LockSettings([&](util::Settings& settings) { settings.rw_settings["name"] = "value"; });    
```
```suggestion
    args1.LockSettings([&](util::Settings& settings) { settings.rw_settings["name"] = "value"; });
```

kiminuo replied:
Will remove to fix linter.

kiminuo committed:
Modify "util_datadir" unit test to not use gArgs.

kiminuo committed:
Change ClearDataDirPathCache() to ArgsManager.ClearPathCache().

kiminuo committed:
scripted-diff: Modify unit tests to use the ArgsManager in the BasicTestingSetup class instead of implicitly relying on gArgs.

-BEGIN VERIFY SCRIPT-
git ls-files src/test/dbwrapper_tests.cpp src/test/denialofservice_tests.cpp src/test/flatfile_tests.cpp src/test/fs_tests.cpp src/test/settings_tests.cpp src/test/util_tests.cpp | xargs sed -i 's/GetDataDir()/m_args.GetDataDirPath()/g';
-END VERIFY SCRIPT-

kiminuo commented on src/util/system.cpp:
```diff
@@ -806,10 +817,7 @@ bool CheckDataDirOption()
-
-    pathCached = fs::path();
-    pathCachedNetSpecific = fs::path();
+    gArgs.ClearDatadirPathCache();
```
bug: `LOCK(csPathCached);` should be here.

kiminuo commented:
@laanwj @hebasto @MarcoFalke @jonatack Friendly ping: Could anyone please tell me if you find the approach to be correct or not?

MarcoFalke commented:
It seems fine to remove a global from a function that should be a pure utility helper

kiminuo commented:
@MarcoFalke So are you saying that my actual approach is good or not? I honestly can't say.

MarcoFalke commented:
Concept ACK, haven't reviewed in detail

ryanofsky commented on src/test/util_tests.cpp:
```diff
@@ -49,24 +49,27 @@ BOOST_FIXTURE_TEST_SUITE(util_tests, BasicTestingSetup)
 {
-    ClearDatadirCache();
-    const fs::path dd_norm = GetDataDir();
+    ArgsManager argsManager;
```
In commit "Modify "util_datadir" unit test to not use gArgs." (ec7769315f5ed62e6145492977226925e122292c)

It seems unnecessary to add another argsmanager for this test, with 3 side-by-argsmanagers:

1. gArgs
2. BasicTestingSetup::m_args_manager added in d4f86a3f608c98365b537d78d7ef1693463d7b9b
3. This local argsManager variable.

Maybe it could just use the basictesting member. If not though, it'd be helpful to have a comment explaining its purpose.

Also see naming in https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md but camelCase should not be used for variable names. If there needs to be a variable would suggest just calling it `args` which is convention for other ArgsManager instances.

kiminuo replied:
The main reason I did it this way was/is that I persuades me that it does what one expects and there is no possible confusion. When I put `gArgs` (or `BasicTestingSetup::m_args_manager`) there, it forces to me wonder: How is `gArgs` actually initialized and where it is? Can some later refactoring affect the test silently? 

Maybe I see it this way because I don't read Bitcoin Core source code for years as other people here. Anyway, I'm open to suggestions, if you still find it better to change the code, I can do it.

_______
Note: An alternative would be to move the test somewhere where there is no `BasicTestingSetup` and no `gArgs` but I'm not sure if there are such tests.

kiminuo replied:
Added comment, thanks.

ryanofsky commented on src/util/system.h:
```diff
@@ -195,6 +195,7 @@ class ArgsManager
     std::map<OptionsCategory, std::map<std::string, Arg>> m_available_args GUARDED_BY(cs_args);
     bool m_accept_any_command GUARDED_BY(cs_args){true};
     std::list<SectionInfo> m_config_sections GUARDED_BY(cs_args);
+    fs::path blocksPathCachedNetSpecific GUARDED_BY(cs_args);
```
In commit "Change GetBlocksDir() to ArgsManager.GetBlocksDirPath()." (fa0b2291aa84fe372bb37a9fcbf1ad3165f7daea)

Again should follow naming convention, would suggest `m_cached_blocks_path`. See naming in  https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md

kiminuo replied:
Applied, thanks.

ryanofsky commented on src/util/system.cpp:
```diff
@@ -723,22 +736,7 @@ fs::path GetDefaultDataDir()
-} // namespace
-
 static fs::path g_blocks_path_cache_net_specific;
-static fs::path pathCached;
-static fs::path pathCachedNetSpecific;
```
In commit "Move StripRedundantLastElementsOfPath before ArgsManager class." (c44993ad96d649be48c94d4ae38d354d609a7282)

This commit seems not to compile because these variables are removed

```
  CXX      libbitcoin_server_a-net.o
util/system.cpp:771:37: error: use of undeclared identifier 'pathCachedNetSpecific'
    fs::path &path = fNetSpecific ? pathCachedNetSpecific : pathCached;
                                    ^
util/system.cpp:771:61: error: use of undeclared identifier 'pathCached'
    fs::path &path = fNetSpecific ? pathCachedNetSpecific : pathCached;
                                                            ^
util/system.cpp:809:5: error: use of undeclared identifier 'pathCached'; did you mean 'csPathCached'?
    pathCached = fs::path();
    ^~~~~~~~~~
    csPathCached
util/system.cpp:740:23: note: 'csPathCached' declared here
static RecursiveMutex csPathCached;
                      ^
util/system.cpp:809:16: error: no viable overloaded '='
    pathCached = fs::path();
    ~~~~~~~~~~ ^ ~~~~~~~~~~
./sync.h:89:16: note: candidate function (the implicit copy assignment operator) not viable: no known conversion from 'fs::path' to 'const AnnotatedMixin<std::recursive_mutex>' for 1st argument
class LOCKABLE AnnotatedMixin : public PARENT
               ^
util/system.cpp:810:5: error: use of undeclared identifier 'pathCachedNetSpecific'
    pathCachedNetSpecific = fs::path();
    ^
```

kiminuo replied:
Fixed, thanks.

ryanofsky commented on src/util/system.cpp:
```diff
@@ -388,6 +388,45 @@ std::optional<unsigned int> ArgsManager::GetArgFlags(const std::string& name) co
+    return path;
+}
+
+void ArgsManager::ClearDatadirPathCache()
```
In commit "Move GetDataDir(fNetSpecific) implementation to ArgsManager." (839324993aeb8448af5a811b7323fa460ac9eac7)

This commit seems not to compile

```
util/system.cpp:422:19: error: out-of-line definition of 'ClearDatadirPathCache' does not match any declaration in 'ArgsManager'; did you mean 'ClearPathCache'?
void ArgsManager::ClearDatadirPathCache()
                  ^~~~~~~~~~~~~~~~~~~~~
                  ClearPathCache                                                                                                                                          
./util/system.h:280:10: note: 'ClearPathCache' declared here
    void ClearPathCache();
         ^
util/system.cpp:820:11: error: no member named 'ClearDatadirPathCache' in 'ArgsManager'
    gArgs.ClearDatadirPathCache();
    ~~~~~ ^
2 errors generated.
```

kiminuo replied:
This is fixed on my system now. Thanks.

ryanofsky commented on src/test/util/setup_common.h:
```diff
@@ -80,6 +81,7 @@ struct BasicTestingSetup {
     ~BasicTestingSetup();
 
     const fs::path m_path_root;
+    ArgsManager m_args_manager;
```
In commit "BasicTestingSetup: Add ArgsManager." (d4f86a3f608c98365b537d78d7ef1693463d7b9b)

Convention other places is to call `ArgsManager` instances `args`, suggest naming this `m_args`

kiminuo replied:
Applied, thanks.

ryanofsky reviewed (commented):
This basically looks good. Other reviewers might be hesitant about this PR adding many new `gArgs` references when it claims to help "Remove gArgs". I think this is fine, because these are just places where `gArgs` was being used implicitly. An alternative to avoid churn would be to pass `ArgsManager&` references functions to places needing it in advance, but if the options for removing `gArgs` are:

1. One big PR removing gArgs
2. Smaller PRs removing gArgs with less churn, but less obvious changes.
3. Smaller PRs removing gArgs with more churn, but more obvious changes.

This PR is choosing option 3.

---

I added some review comments below. Suggestions are:

- Fix compile errors in intermediate commits
- For consistency and brevity call `ArgsManager` instances `args` or `m_args` not `argsManager` or `m_args_manager`
- Fix other camelCase names

ryanofsky commented:
The new `BasicTestingSetup` args manager instance doesn't seem to be used very much right now. I'd think here or at some point in the future it'd make sense for `BasicTestingSetup::m_node.args` to be a pointer to the local test args manager instead of to `gArgs`. Maybe a change like the following would make sense:

```diff
diff --git a/src/bitcoind.cpp b/src/bitcoind.cpp
index 80ab69c131e..6fddb469e56 100644
--- a/src/bitcoind.cpp
+++ b/src/bitcoind.cpp
@@ -113,8 +113,8 @@ static bool AppInit(int argc, char* argv[])
     util::ThreadSetInternalName("init");
 
     // If Qt is used, parameters/bitcoin.conf are parsed in qt/bitcoin.cpp's main()
-    SetupServerArgs(node);
     ArgsManager& args = *Assert(node.args);
+    SetupServerArgs(args);
     std::string error;
     if (!args.ParseParameters(argc, argv, error)) {
         return InitError(Untranslated(strprintf("Error parsing command line arguments: %s\n", error)));
diff --git a/src/init.cpp b/src/init.cpp
index 0d44b75b727..7b024e14686 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -358,12 +358,8 @@ static void OnRPCStopped()
     LogPrint(BCLog::RPC, "RPC stopped.\n");
 }
 
-void SetupServerArgs(NodeContext& node)
+void SetupServerArgs(ArgsManager& argsman)
 {
-    assert(!node.args);
-    node.args = &gArgs;
-    ArgsManager& argsman = *node.args;
-
     SetupHelpOptions(argsman);
     argsman.AddArg("-help-debug", "Print help message with debugging options and exit", ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST); // server-only for now
 
diff --git a/src/init.h b/src/init.h
index 5d01d4c1ac3..98b83f0cf7f 100644
--- a/src/init.h
+++ b/src/init.h
@@ -69,7 +69,7 @@ bool AppInitMain(const std::any& context, NodeContext& node, interfaces::BlockAn
 /**
  * Register all arguments with the ArgsManager
  */
-void SetupServerArgs(NodeContext& node);
+void SetupServerArgs(ArgsManager& argsman);
 
 /** Returns licensing information (for -version) */
 std::string LicenseInfo();
diff --git a/src/qt/bitcoin.cpp b/src/qt/bitcoin.cpp
index fc6d0febc2d..7eeeb3c1155 100644
--- a/src/qt/bitcoin.cpp
+++ b/src/qt/bitcoin.cpp
@@ -477,7 +477,7 @@ int GuiMain(int argc, char* argv[])
 
     /// 2. Parse command-line options. We do this after qt in order to show an error if there are problems parsing these
     // Command-line options take precedence:
-    SetupServerArgs(node_context);
+    SetupServerArgs(gArgs);
     SetupUIArgs(gArgs);
     std::string error;
     if (!gArgs.ParseParameters(argc, argv, error)) {
diff --git a/src/test/util/setup_common.cpp b/src/test/util/setup_common.cpp
index 1c961be419d..a460eab906d 100644
--- a/src/test/util/setup_common.cpp
+++ b/src/test/util/setup_common.cpp
@@ -74,6 +74,7 @@ BasicTestingSetup::BasicTestingSetup(const std::string& chainName, const std::ve
     : m_path_root{fs::temp_directory_path() / "test_common_" PACKAGE_NAME / g_insecure_rand_ctx_temp_path.rand256().ToString()},
       m_args_manager{}
 {
+    m_node.args = &m_args_manager;
     const std::vector<const char*> arguments = Cat(
         {
             "dummy",
@@ -92,7 +93,7 @@ BasicTestingSetup::BasicTestingSetup(const std::string& chainName, const std::ve
     gArgs.ForceSetArg("-datadir", m_path_root.string());
     gArgs.ClearPathCache();
     {
-        SetupServerArgs(m_node);
+        SetupServerArgs(*m_node.args);
         std::string error;
         const bool success{m_node.args->ParseParameters(arguments.size(), arguments.data(), error)};
         assert(success);
```

kiminuo commented:
@ryanofsky Thank you for the review! I will try to address it soon.

kiminuo committed:
Move StripRedundantLastElementsOfPath before ArgsManager class.

kiminuo committed:
Move GetDataDir(fNetSpecific) implementation to ArgsManager.

kiminuo committed:
Change GetBlocksDir() to ArgsManager.GetBlocksDirPath().

kiminuo commented:
> The new `BasicTestingSetup` args manager instance doesn't seem to be used very much right now. I'd think here or at some point in the future it'd make sense for `BasicTestingSetup::m_node.args` to be a pointer to the local test args manager instead of to `gArgs`. Maybe a change like the following would make sense:

I would like to reserve this for a follow-up PR. Other than that, I have attempted to address all your suggestions. Notably, I tried to run `make check` for each commit of this PR. I can try that again in the evening with clear head.

kiminuo commented on src/util/system.h:
```diff
@@ -200,6 +200,8 @@ class ArgsManager
     bool m_accept_any_command GUARDED_BY(cs_args){true};
     std::list<SectionInfo> m_config_sections GUARDED_BY(cs_args);
+    fs::path datadirPathCached GUARDED_BY(cs_args);
+    fs::path datadirPathCachedNetSpecific GUARDED_BY(cs_args);
```
This is fixed on my system now. Thanks.

kiminuo replied:
Applied, thanks.

kiminuo commented:
"Cirrus CI / [depends, sanitizers: thread (TSan), no gui] [focal]" seems to be flaky lately.

jonatack commented:
> "Cirrus CI / [depends, sanitizers: thread (TSan), no gui] [focal]" seems to be flaky lately.

If you rebase on latest master I think that solves it.

kiminuo commented:
> If you rebase on latest master I think that solves it.

Let's try it :)

ryanofsky commented on src/util/system.cpp:
```diff
@@ -737,8 +776,6 @@ fs::path GetDefaultDataDir()
 
 static fs::path g_blocks_path_cache_net_specific;
-static fs::path pathCached;
-static fs::path pathCachedNetSpecific;
```
In commit "Move GetDataDir(fNetSpecific) implementation to ArgsManager." (fac964698fab1b269d4414dc54560b29a3eccd18)

Note (no change suggested): I was initially confused why `pathCached` and `pathCachedNetSpecific` were being removed in this commit without also removing `g_blocks_path_cache_net_specific`. But it's removed in a later commit, and the reason for treating it differently just seems to be that it's needed many fewer places, so the `::GetBlocksDir` wrapper function can be dropped right away, unlike `::GetDataDir`

kiminuo replied:
My primary objective was to tackle `::GetDataDir` but then it led me to handle `::GetBlocksDir` too. I did not want to modify those two at the same time.

ryanofsky commented on src/test/util_tests.cpp:
```diff
@@ -49,24 +49,28 @@ BOOST_FIXTURE_TEST_SUITE(util_tests, BasicTestingSetup)
-    BOOST_CHECK_EQUAL(dd_norm, GetDataDir());
-
-    gArgs.ForceSetArg("-datadir", dd_norm.string() + "/.//");
-    ClearDatadirCache();
```
In commit "Modify "util_datadir" unit test to not use gArgs." (8584666393255a7ce62d741ddd8d994c486423c6)

It seems like the checks in this test might be meaningless now that the ClearDatadirCache calls are dropped, because the repeated `BOOST_CHECK_EQUAL(dd_norm, args.GetDataDirPath());` checks will keep passing no matter what value `-datadir` is force set to.

I think it would be best for this commit to either keep the Clear calls so the test is changed as little as possible. Or the test could be rewritten to use different expected values or different args objects.

kiminuo replied:
You are right. I have put those commands back.

ryanofsky reviewed (commented):
Code review almost-ACK 92b1890c9eeeecf9cf7325ef3dc1d2c44a588b87, modulo the `util_datadir` commit 8584666393255a7ce62d741ddd8d994c486423c6, which I think breaks that test in a way where it still passes but no longer meaningfully checks things (see suggestion below).

I would also consider dropping the 6e7e67f855282ae4baa3756f2e95aee12de13f91 commit from this PR because I think it makes the PR a lot bigger, causes code churn, and doesn't add as much value as the other changes here. 6e7e67f855282ae4baa3756f2e95aee12de13f91 could be its own separate followup PR. (@dongcarl could be good person to ask about it, too.)

Otherwise everything looks great.

kiminuo commented:
> Code review almost-ACK [92b1890](https://github.com/bitcoin/bitcoin/commit/92b1890c9eeeecf9cf7325ef3dc1d2c44a588b87), modulo the `util_datadir` commit [8584666](https://github.com/bitcoin/bitcoin/commit/8584666393255a7ce62d741ddd8d994c486423c6), which I think breaks that test in a way where it still passes but no longer meaningfully checks things (see suggestion below).

I have put `ClearDatadirCache` calls back. This was an unintended change I missed.

> I would also consider dropping the [6e7e67f](https://github.com/bitcoin/bitcoin/commit/6e7e67f855282ae4baa3756f2e95aee12de13f91) commit from this PR because I think it makes the PR a lot bigger, causes code churn, and doesn't add as much value as the other changes here. [6e7e67f](https://github.com/bitcoin/bitcoin/commit/6e7e67f855282ae4baa3756f2e95aee12de13f91) could be its own separate followup PR. (@dongcarl could be good person to ask about it, too.)

You are right, the commit makes the PR too large probably. I have dropped the 6e7e67f (Change GetDataDir() calls to gArgs.GetDataDirPath() calls.) commit.

ryanofsky reviewed (approved):
Code review ACK 834b971bff371951e3e63617d0a3fb9b3ad79c62. Thanks for fast updates! Only change last review was restoring util_datadir test, and dropping a commit so the PR is now limited to src/util/system (and tests) and doesn't affect outside code.

ryanofsky commented:
Noticed this in IRC:

> \<Kiminuo> Hi guys, a quick question: I have this PR https://github.com/bitcoin/bitcoin/pull/21244 and it got one ACK. Now, should I just wait patiently until somebody else will review the PR? Or am I supposed to do something else to increase the chance of merging it? I'm not in hurry. I would just like to understand the review process better.

This is an easy review that mostly affects tests so I'd encourage anybody who cares about c++ unit tests to look at this PR.

Also for Kiminuo, I think a good way to get other people to review your PRs is to review their PRs. Progress on bitcoin is limited more by lack of review than lack of code contributions, and reviews that are substantive and show thought & understanding are encouraged even if you don't have prior experience reviewing bitcoin PRs (https://bitcoincore.reviews/ is a good resource for this).

jonatack commented:
> I think a good way to get other people to review your PRs is to review their PRs. Progress on bitcoin is limited more by lack of review than lack of code contributions, and reviews that are substantive and show thought & understanding are encouraged even if you don't have prior experience reviewing bitcoin PRs (https://bitcoincore.reviews/ is a good resource for this).

Agree! suggested a similar approach to Kiminuo yesterday via IRC DM.

hebasto commented on src/util/system.h:
```diff
@@ -200,6 +196,9 @@ class ArgsManager
     std::map<OptionsCategory, std::map<std::string, Arg>> m_available_args GUARDED_BY(cs_args);
     bool m_accept_any_command GUARDED_BY(cs_args){true};
     std::list<SectionInfo> m_config_sections GUARDED_BY(cs_args);
+    fs::path m_cached_blocks_path GUARDED_BY(cs_args);
+    fs::path m_cached_datadir_path GUARDED_BY(cs_args);
+    fs::path m_cached_network_datadir_path GUARDED_BY(cs_args);
```
As these data members are cache, why do not declare them `mutable`?
It will allow to keep `const`ness of the `GetSettingsPath` and `WriteSettingsFile`.

kiminuo replied:
I have modified this.

hebasto commented on src/util/system.h:
```diff
@@ -263,6 +262,27 @@ class ArgsManager
+     */
+    const fs::path &GetDataDirPath(bool fNetSpecific = true);
+
+    /**
+     * For testing
+     */
+    void ClearPathCache();
```
The description of this function is not true: https://github.com/bitcoin/bitcoin/blob/834b971bff371951e3e63617d0a3fb9b3ad79c62/src/util/system.cpp#L974

kiminuo replied:
Do you have a suggestion for the description?

I think the comment is mostly unhelpful as it is too generic anyway. I would write something like that it's useful to propagate changes in args. I'm not sure about proper English wording though.

hebasto replied:
Obviously, I'm not an [Official Namer Of Stuff](https://github.com/bitcoin/bitcoin/pull/20861#discussion_r596389178) :)

Maybe "Clean cached directory paths"?

kiminuo replied:
Will be addressed in the next push.

hebasto commented on src/util/system.h:
```diff
@@ -263,6 +262,27 @@ class ArgsManager
+     * @return Absolute path on success, otherwise an empty path when a non-directory path would be returned
+     * @post Returned directory path is created unless it is empty
+     */
+    const fs::path &GetDataDirPath(bool fNetSpecific = true);
```
```suggestion
    const fs::path& GetDataDirPath(bool net_specific = true);
```

kiminuo replied:
Yes, thank you.

hebasto commented on src/util/system.cpp:
```diff
@@ -375,6 +388,72 @@ std::optional<unsigned int> ArgsManager::GetArgFlags(const std::string& name) co
+const fs::path& ArgsManager::GetDataDirPath(bool fNetSpecific)
+{
+    LOCK(cs_args);
+    fs::path &path = fNetSpecific ? m_cached_network_datadir_path : m_cached_datadir_path;
```
style nit:
```suggestion
    fs::path& path = fNetSpecific ? m_cached_network_datadir_path : m_cached_datadir_path;
```

kiminuo replied:
Yes, thank you.

hebasto commented on src/util/system.cpp:
```diff
@@ -375,6 +388,72 @@ std::optional<unsigned int> ArgsManager::GetArgFlags(const std::string& name) co
+const fs::path& ArgsManager::GetBlocksDirPath()
+{
+    LOCK(cs_args);
+    fs::path &path = m_cached_blocks_path;
```
style nit:
```suggestion
    fs::path& path = m_cached_blocks_path;
```

hebasto reviewed (commented):
Approach ACK 834b971bff371951e3e63617d0a3fb9b3ad79c62.

I see no reasons to guard new data members `m_cached_*_path` with `RecursiveMutex cs_args`. It tangles recursive locking of `cs_args` more and more. We should move from `RecursiveMutex` to `Mutex` (#19303, #19213). To not deteriorate status quo, maybe just use a new specific mutex?

kiminuo commented:
> I see no reasons to guard new data members `m_cached_*_path` with `RecursiveMutex cs_args`. It tangles recursive locking of `cs_args` more and more. We should move from `RecursiveMutex` to `Mutex` (#19303, #19213). To not deteriorate status quo, maybe just use a new specific mutex?

Would it be better to do this in a follow-up PR?

hebasto commented:
> > I see no reasons to guard new data members `m_cached_*_path` with `RecursiveMutex cs_args`. It tangles recursive locking of `cs_args` more and more. We should move from `RecursiveMutex` to `Mutex` (#19303, #19213). To not deteriorate status quo, maybe just use a new specific mutex?
> 
> Would it be better to do this in a follow-up PR?

Not sure if it is better, but it won't stop me from ACKing :)

hebasto commented:
- [x] 70cdf679f8e665dbdc3301873a0267fe9faa72cd - "Move StripRedundantLastElementsOfPath before ArgsManager class."
- [ ] 1add318704108faa98f5b1b8e9c96d960e9d23a8 - "Move GetDataDir(fNetSpecific) implementation to ArgsManager."
- [ ] 5441bd3f54e675445f5a9c10462c24fbba721243 - "Modify "util_datadir" unit test to not use gArgs."
- [ ] ea70f2266b4329cce75ccb8db73e426055f345a8 - "BasicTestingSetup: Add ArgsManager."
- [ ] 46c8546e4d42b4825b82adf3fd1ff42591850530 - "Modify unit tests to use the ArgsManager in the BasicTestingSetup class instead of implicitly relying on gArgs."
- [ ] 3cf594e1dc1131f6e39911e8ddc5721f310e5aeb - "Change GetBlocksDir() to ArgsManager.GetBlocksDirPath()."
- [ ] 69b8c4ad40185862fe199005c81a75d582053bd4 - "Change ClearDataDirPathCache() to ArgsManager.ClearPathCache()."

hebasto commented on src/test/util_tests.cpp:
```diff
@@ -49,24 +49,27 @@ BOOST_FIXTURE_TEST_SUITE(util_tests, BasicTestingSetup)
-    gArgs.ForceSetArg("-datadir", dd_norm.string() + "/");
-    ClearDatadirCache();
-    BOOST_CHECK_EQUAL(dd_norm, GetDataDir());
+    fs::path dd_norm = args.GetDataDirPath();
```
Why `const` removed?

kiminuo replied:
Will be addressed in the next push. Thanks.

hebasto reviewed (approved):
ACK 69b8c4ad40185862fe199005c81a75d582053bd4, I have reviewed the code and it looks OK, tested locally.

nit: It seems 46c8546e4d42b4825b82adf3fd1ff42591850530 could be a scripted-diff.

Please next time do not rebase without necessity :)

kiminuo commented:
> Please next time do not rebase without necessity :)

Interesting point, it did not occur to me that it's probably better. Thank you for pointing out.

MarcoFalke commented on src/test/util/setup_common.cpp:
```diff
@@ -87,6 +88,7 @@ BasicTestingSetup::BasicTestingSetup(const std::string& chainName, const std::ve
         extra_args);
     util::ThreadRename("test");
     fs::create_directories(m_path_root);
+    m_args.ForceSetArg("-datadir", m_path_root.string());
```
There is already m_node.args. Any reason to avoid using that?

ryanofsky replied:
> There is already m_node.args. Any reason to avoid using that?

I suggested this for a followup (It would be a bigger change involving init and might also require updating more tests). Diff was here: https://github.com/bitcoin/bitcoin/pull/21244#issuecomment-816026224

MarcoFalke replied:
Until then I'd prefer to use `m_node.args->` (or `gArgs.` if that is too much typing). It will be a trivial scripted diff to change that to `m_args.`. But having three argsman in the tests (two of which are the same and one that is different) will just cause confusion and makes review harder/code more fragile. How would you protect against their state diverging?

ryanofsky replied:
I'm happy with any intermediate state. Here's what I think the ultimate goal should be:

- Tests should have an `ArgsManager m_args` member that is newly initialized before each test and completely destroyed after each test.
- Where possible, tests should pass `m_args` references directly to functions that are being tested and need to read options.
- Tests can pass `m_args` pointers indirectly through `NodeContext` when they are calling node RPC methods, or through `WalletContext` when they are calling wallet RPC methods.
- Unless tests are calling RPC methods or init code, they probably don't need `NodeContext` and `WalletContext` instances and it would be nice to drop `NodeContext` from the base test class.
- `gArgs` should go away in node and wallet code. It might be useful to keep `gArgs` in test code to read test options (random seeds, timeouts, log options, debug options). It might make sense to keep `gArgs` in GUI code since GUI code relies on other global singletons like `QApplications` and `QSettings` and trying to get rid of it might just add extra complexity.

The current PR seems just like a good a first step in this direction to me since it adds local test `ArgsManager` instances and starts to use them minimally.

These are just my thoughts though. I think any approach you want to take here is probably reasonable.

MarcoFalke replied:
Fair enough, any approach fine by me, too

kiminuo replied:
I think Russell summarized it very nicely. 

My approach with this PR is to use `m_args` wrt `GetDataDir()` functionality for tests. Yes, it is somewhat fragile at this point. If you are too concerned about this, I can revert that change in `BasicTestingSetup` and leave it for some follow-up PR. 

note: There are tests that just need `GetDataDir()` function and not anything else from `ArgsManager`, these tests benefit from using `m_args` (instead of `gArgs`).

note 2: Originally, https://github.com/bitcoin/bitcoin/commit/6e7e67f855282ae4baa3756f2e95aee12de13f91 was a part of this PR and with this, one can see quite nicely where `gArgs` is used and that's one step closer to remove it from those places and make some more tests to be true unit tests.

MarcoFalke commented on src/test/getarg_tests.cpp:
```diff
@@ -172,24 +172,24 @@ BOOST_AUTO_TEST_CASE(boolargno)
-    BOOST_CHECK(m_args.GetBoolArg("-foo", true));
-    BOOST_CHECK(m_args.GetBoolArg("-foo", false));
+    BOOST_CHECK(m_local_args.GetBoolArg("-foo", true));
+    BOOST_CHECK(m_local_args.GetBoolArg("-foo", false));
```
This diff could be a scripted-diff in a separate commit?

kiminuo replied:
Yes, will be addressed in the next push.

kiminuo committed:
BasicTestingSetup: Add ArgsManager.

kiminuo committed:
scripted-diff: Replace m_args with m_local_args in getarg_tests.cpp

-BEGIN VERIFY SCRIPT-
git ls-files src/test/getarg_tests.cpp | xargs sed -i "s/m_args/m_local_args/g";
-END VERIFY SCRIPT-

kiminuo commented:
I addessed outstanding suggestions. Diff since 69b8c4ad is minimal.

ryanofsky reviewed (approved):
Code review ACK bb8d1c6e029a2fd01387599d2ff3bfc879a8ada5. Just minor const/naming changes and splitting/scripting commits since last review

maflcko reviewed (commented):
review ACK bb8d1c6e029a2fd01387599d2ff3bfc879a8ada5 📓

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK bb8d1c6e029a2fd01387599d2ff3bfc879a8ada5 📓
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYEACgkQzit1aX5p
pUgg4wv/VxpdghXhej+GNGTyDSSqYsXme76aDsRmpsxCgFsPO1mpnR9KzmCQNYN7
M9MBJ8+1Slsa8an8P5pk9cFYmGciu0Iu26hVOrZ/lESZYZ4j9+qWbN5dN2VkSDEg
GEt1P1RnE5qyqLQZmrrafkk0EcGl8ErBc56kjlYyciugjKX8KNqdvDSJAAdnMCmB
h+TIlGuCXrdG2TEngv0neHJ5nrp4JG0hdkD73u8QeVnszs3KuHbXzWenng5jJKwC
CidMrDLjLtc/B/I31GlXeWD3pYwlOkxrtC5rAObaThRtU0NlQ5Pz3DsrcvWLuEPA
vzyBa/ru6TUHxGPpq4IMTjdf0v3GVYyjJc0xbXiGGLQygJw6ff0YzJm8F0hqw7J+
/6xvt8hVz+10z887VNIBQLKgasLiDB5R1us+XprxYcwGW3GljiDtKCFUcIBvSlbn
D3yODkhg6jYIoiLUi2xp8B2e9AWqlCSx09dySUl3Gn+Q9srvkHIu5ZOfUg1vM6rx
XhwN4wTz
=q1yj
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `977786d72da2ca0c8d5b6be0fa8686d3e4ab5578eca910bb84af1e639ee23595  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108977786d72da2ca0c8d5b6be0fa8686d3e4ab5578eca910bb84af1e639ee23595f010730cfa5fac2ce2e82c6fc6389bcfac8208fff010059aa27dae109acb985b35b561faac1c08f104607dd280f0089460d533ad66e9cd0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01037b19f12b960ae34908982443ed7e2ae08f104607dc4a5f008052bfd24d668e4f70083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0107e2c1648636f4db3e3a514765c2b94e608f104607dd27ff008528e10b24b05da740083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f0108484ce6fab79a566307aeb11ba74781008f104607dd280f008ea8530457a596b510083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

hebasto reviewed (approved):
re-ACK bb8d1c6e029a2fd01387599d2ff3bfc879a8ada5, addressed comments, and two commits made scripted-diffs since my [previous](https://github.com/bitcoin/bitcoin/pull/21244#pullrequestreview-638270583) review.
